include/dice/template-library/integral_template_variant_v2.hpp

```diff
@@ -0,0 +1,153 @@
+	 *
+	 * Direction is automatic:
+	 * - first < last: ascending [first, last) → T<first> .. T<last-1>
+	 * - first > last: descending (last, first] → T<first> .. T<last+1>
+	 * - first == last: std::variant<std::monostate> (empty)
+	 *
+	 * @tparam first the first ix for T<ix> (inclusive)
+	 * @tparam last the last ix (exclusive, not included in the variant)
+	 * @tparam T the template that gets instantiated with T<ix>
+	 */
+	template<std::integral auto first, decltype(first) last, template<decltype(first)> typename T>
+	struct integral_template_variant_v2 {
+		using index_type = decltype(first);
+		using underlying_type = itv_detail_v2::make_variant<first, last, T>;
+
+		template<index_type ix>
+		using value_type = T<ix>;
+
+	private:
+		template<index_type ix>
+		static constexpr bool check_ix_v = integral_sequence::valid_index_v<first, last, ix>;
+
+		underlying_type repr_;
+
+	public:
+		constexpr integral_template_variant_v2() noexcept(std::is_nothrow_default_constructible_v<underlying_type>) = default;
+		constexpr integral_template_variant_v2(integral_template_variant_v2 const &other) noexcept(std::is_copy_constructible_v<underlying_type>) = default;
+		constexpr integral_template_variant_v2(integral_template_variant_v2 &&other) noexcept(std::is_nothrow_move_constructible_v<underlying_type>) = default;
+		constexpr integral_template_variant_v2 &operator=(integral_template_variant_v2 const &other) noexcept(std::is_nothrow_copy_assignable_v<underlying_type>) = default;
+		constexpr integral_template_variant_v2 &operator=(integral_template_variant_v2 &&other) noexcept(std::is_nothrow_move_assignable_v<underlying_type>) = default;
+		constexpr ~integral_template_variant_v2() noexcept(std::is_nothrow_destructible_v<underlying_type>) = default;
+
+		template<index_type ix>
+		constexpr integral_template_variant_v2(T<ix> const &value) noexcept(std::is_nothrow_copy_constructible_v<T<ix>>)
+			: repr_{value} {
+			(void) check_ix_v<ix>;
+		}
+
+		template<index_type ix>
+		constexpr integral_template_variant_v2(T<ix> &&value) noexcept(std::is_nothrow_move_constructible_v<T<ix>>)
+			: repr_{std::move(value)} {
+			(void) check_ix_v<ix>;
+		}
+
+		template<typename U, typename... Args>
+		explicit constexpr integral_template_variant_v2(std::in_place_type_t<U>, Args &&...args) noexcept(std::is_nothrow_constructible_v<U, decltype(std::forward<Args>(args))...>)
+			: repr_{std::in_place_type<U>, std::forward<Args>(args)...} {
+		}
+
+		template<index_type ix, typename... Args>
+		constexpr explicit integral_template_variant_v2(std::in_place_index_t<ix>, Args &&...args)
+			: repr_(std::in_place_type<T<ix>>, std::forward<Args>(args)...) {
+			(void) check_ix_v<ix>;
+		}
+
+		template<index_type ix, typename... Args>
+		constexpr value_type<ix> &emplace(Args &&...args) {
+			(void) check_ix_v<ix>;
+			return repr_.template emplace<T<ix>>(std::forward<Args>(args)...);
+		}
+
+		[[nodiscard]] static constexpr size_t size() noexcept {
```

Comment:
I would remove this function, because size implies that there are multiple elements present.